Add workaround to fix the modified state of rancher-monitoring-crd (#541)
harvester/docs: docs/troubleshooting/monitoring.md

    rancher-monitoring-prometheus-adapter-55dc9ccd5d-vtssc   1/1     Running     0          3m55s
    rancher-monitoring-prometheus-node-exporter-lgb88        1/1     Running     0          3m55s
    ```

> ## The state of the rancher-monitoring-crd Managedchart is modified

Reviewer: ```suggestion
## `rancher-monitoring-crd` ManagedChart State is `Modified`
```

### Issue Description

> Occasionally the state of the `rancher-monitoring-crd` ManagedChart is `Modified`, and the message is like `rancher-monitoring-crd-manager missing`.

Reviewer: ```suggestion
In certain situations, the state of the `rancher-monitoring-crd` ManagedChart object changes to `Modified` (with the message `rancher-monitoring-crd-manager missing`).

Example:
```

apiVersion: management.cattle.io/v3
kind: ManagedChart
...
spec:
  chart: rancher-monitoring-crd
  defaultNamespace: cattle-monitoring-system
  paused: false
  releaseName: rancher-monitoring-crd
  repoName: harvester-charts
  targets:
  - clusterName: local
    clusterSelector:
      matchExpressions:
      - key: provisioning.cattle.io/unmanaged-system-agent
        operator: DoesNotExist
  version: 102.0.0+up40.1.2
...
status:
  conditions:
  - lastUpdateTime: "2024-02-22T14:03:11Z"
    message: Modified(1) [Cluster fleet-local/local]; clusterrole.rbac.authorization.k8s.io
      rancher-monitoring-crd-manager missing; clusterrolebinding.rbac.authorization.k8s.io
      rancher-monitoring-crd-manager missing; configmap.v1 cattle-monitoring-system/rancher-monitoring-crd-manifest
      missing; serviceaccount.v1 cattle-monitoring-system/rancher-monitoring-crd-manager
      missing
    status: "False"
    type: Ready
  - lastUpdateTime: "2024-02-22T14:03:11Z"
    status: "True"
    type: Processed
  - lastUpdateTime: "2024-04-02T07:45:26Z"
    status: "True"
    type: Defined
  display:
    readyClusters: 0/1
    state: Modified
...
```

> The `Managedchart` object has a downstream object `Bundle`, which also has similar information

Reviewer: ```suggestion
The `ManagedChart` object has a downstream object named `Bundle`, which has similar information.
```

```
$ kubectl get bundles -A
NAMESPACE     NAME                                          BUNDLEDEPLOYMENTS-READY   STATUS
fleet-local   fleet-agent-local                             1/1
fleet-local   local-managed-system-agent                    1/1
fleet-local   mcc-harvester                                 1/1
fleet-local   mcc-harvester-crd                             1/1
fleet-local   mcc-local-managed-system-upgrade-controller   1/1
fleet-local   mcc-rancher-logging-crd                       1/1
fleet-local   mcc-rancher-monitoring-crd                    0/1                       Modified(1) [Cluster fleet-local/local]; clusterrole.rbac.authorization.k8s.io rancher-monitoring-crd-manager missing; clusterrolebinding.rbac.authorization.k8s.io rancher-monitoring-crd-manager missing; configmap.v1 cattle-monitoring-system/rancher-monitoring-crd-manifest missing; serviceaccount.v1 cattle-monitoring-system/rancher-monitoring-crd-manager missing
```

> When this issue has happend and you [start an upgrade](../upgrade/automatic.md#start-an-upgrade), Harvester may return such an error message: `admission webhook "validator.harvesterhci.io" denied the request: managed chart rancher-monitoring-crd is not ready, please wait for it to be ready`.
> 
> However when you check those `missing` objects, they are existing in the cluster.

Reviewer: ```suggestion
When the issue exists and you [start an upgrade](../upgrade/automatic.md#start-an-upgrade), Harvester may return the following error message: `admission webhook "validator.harvesterhci.io" denied the request: managed chart rancher-monitoring-crd is not ready, please wait for it to be ready`.

Also, when you search for the objects marked as `missing`, you will find that the objects exist in the cluster.

Example:
```

  ...
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: rancher-monitoring-crd-manager
subjects:
- kind: ServiceAccount
  name: rancher-monitoring-crd-manager
  namespace: cattle-monitoring-system

$ kubectl get configmap -n cattle-monitoring-system rancher-monitoring-crd-manifest
apiVersion: v1
data:
  crd-manifest.tgz.b64: ...
kind: ConfigMap
metadata:
  annotations:
    meta.helm.sh/release-name: rancher-monitoring-crd
    meta.helm.sh/release-namespace: cattle-monitoring-system
  creationTimestamp: "2023-01-09T11:04:33Z"
  labels:
    app.kubernetes.io/managed-by: Helm
  name: rancher-monitoring-crd-manifest
  namespace: cattle-monitoring-system
  ...

$ kubectl get ServiceAccount -n cattle-monitoring-system rancher-monitoring-crd-manager
apiVersion: v1
kind: ServiceAccount
metadata:
  annotations:
    meta.helm.sh/release-name: rancher-monitoring-crd
    meta.helm.sh/release-namespace: cattle-monitoring-system
  creationTimestamp: "2023-01-09T11:04:33Z"
  labels:
    app: rancher-monitoring-crd-manager
    app.kubernetes.io/managed-by: Helm
  name: rancher-monitoring-crd-manager
  namespace: cattle-monitoring-system
  ...
```

### Root Cause

> Those objects do not have the related labels/annotations required by the `ManagedChart` object, hence they are marked as `missging`.

Reviewer: ```suggestion
The objects that are marked as `missing` do not have the related labels and annotations required by the `ManagedChart` object.
```

### Workaround

> To fix the issue, you may follow the steps below to re-create the `rancher-monitoring-crd` Managedchart object.
> 
> 1. Confirm the chart version of `rancher-monitoring-crd` from the `harvester-cluster-repo` service.
> 
>     Get the `harvester-cluster-repo` service IP.

Reviewer: ```suggestion
To fix the issue, you must recreate the `rancher-monitoring-crd` Managedchart object.

1. Verify the chart version of `rancher-monitoring-crd` using the `harvester-cluster-repo` service.

    Get the IP of the `harvester-cluster-repo` service.
```

    ```
    $ kubectl get service -A | grep repo
    cattle-system                     harvester-cluster-repo                        ClusterIP      10.53.146.66    <none>            80/TCP                         8d

    The IP address is 10.53.146.66.
    ```

>     Get the chart version from the `index.yaml` in the `harvester-cluster-repo`.

Reviewer 2: It would be thoughtful if we kindly remind the user where to run the `curl` command (I assume it could be run on the Harvester node?)

>     Get the chart version from the `index.yaml` in the `harvester-cluster-repo`.

Reviewer: ```suggestion
    Get the chart version from the `index.yaml` file in `harvester-cluster-repo`.
```

    ```
    $ curl http://10.53.146.66/charts/index.yaml | grep "rancher-monitoring"
    ...
    name: rancher-monitoring
        - rancher-monitoring-102.0.0+up40.1.2.tgz
      rancher-monitoring-crd:
          catalog.cattle.io/release-name: rancher-monitoring-crd
        description: Installs the CRDs for rancher-monitoring.
        name: rancher-monitoring-crd
        - rancher-monitoring-crd-102.0.0+up40.1.2.tgz

    The chart version is 102.0.0+up40.1.2.
    ```

>     The chart version should be the same as the `version` field in the `rancher-monitoring-crd` ManagedChart object.

Reviewer: ```suggestion
    The chart version must be identical to the value of the `version` field in the `rancher-monitoring-crd` ManagedChart object.
```

    :::note

>     If those two versions are different, you need to stop and ask help from Harvester support.

Reviewer: Should all users be encouraged to contact support directly? If so, how exactly are they supposed to do that?

cc: @bk201

Author: Only paid customers can contact Harvester support directly. Community users will ask on github or slack channel.

Reviewer 3: Enterprise users know how to report issues and contact SUSE support. A better approach here might be to create a page like "How to ask for help?" that describes how to get help from Slack/Github. And link to that page. Although I think the work can be done in another PR.

Author: Added.

    :::

> 1. Disable the `rancher-monitoring` [addon](../advanced/addons.md),  wait until its state is `Disabled`.

Reviewer: ```suggestion
1. Disable the `rancher-monitoring` [addon](../advanced/addons.md). Wait until the state changes to `Disabled`.
```

> 1. Delete the `rancher-monitoring-crd` ManagedChart.

Reviewer: ```suggestion
1. Delete the `rancher-monitoring-crd` ManagedChart.
```

   ```
   $ kubectl delete managedchart -n fleet-local rancher-monitoring-crd
   ```

> 1. Delete those marked as `missing` objects manually.

Reviewer: ```suggestion
1. Manually delete all objects marked as `missing`.
```

>     When any of the following objects is found, delete it.

Reviewer: This is redundant. Please remove it.

    ```
    $ kubectl get clusterrole rancher-monitoring-crd-manager
    $ kubectl get clusterrolebinding rancher-monitoring-crd-manager
    $ kubectl get configmap -n cattle-monitoring-system rancher-monitoring-crd-manifest
    $ kubectl get ServiceAccount -n cattle-monitoring-system rancher-monitoring-crd-manager
    ```

> 1. Prepare a new yaml file for the `rancher-monitoring-crd` ManagedChart.
> 
>     Replace the version field with the confirmed value in the first step.

Reviewer: ```suggestion
1. Prepare a new YAML file for the `rancher-monitoring-crd` ManagedChart.

    In the `version` field, specify the chart version from step 1. The chart version must be identical to the value of the `version` field in the `rancher-monitoring-crd` ManagedChart object.
```

    ```
    $ cat > rm-crd.yaml << 'EOF'
    apiVersion: management.cattle.io/v3
    kind: ManagedChart
    metadata:
      name: rancher-monitoring-crd
      namespace: fleet-local
    spec:
      chart: rancher-monitoring-crd
      releaseName: rancher-monitoring-crd
      version: "102.0.0+up40.1.2"
      defaultNamespace: cattle-monitoring-system
      repoName: harvester-charts
      targets:
      - clusterName: local
        clusterSelector:
          matchExpressions:
          - key: provisioning.cattle.io/unmanaged-system-agent
            operator: DoesNotExist
    EOF
    ```

> 1. Create the `rancher-monitoring-crd` ManagedChart object.

Reviewer: ```suggestion
1. Create the `rancher-monitoring-crd` ManagedChart object.
```

    ```
    $ kubectl apply -f rm-crd.yaml
    ```

> 1. Wait until the new ManagedChart object is ready. The state of the rancher-monitoring-crd Managedchart will not be `modified` any more.

Reviewer: ```suggestion
1. Verify that the new `rancher-monitoring-crd` ManagedChart object was created. The state of the object should not be `modified`.
```

    ```
    $ kubectl get managedchart -A
    ```

> 1. Enable the `rancher-monitoring` [addon](../advanced/addons.md), wait until its state is `DeploySuccessful`.

Reviewer: ```suggestion
1. Enable the `rancher-monitoring` [addon](../advanced/addons.md). Wait until the state changes to `DeploySuccessful`.
```

> 1. If your previous upgrade operation was denied due to this issue, now you can retry the upgrade.

Reviewer: ```suggestion
1. (Optional) Retry the upgrade (if previously unsuccessful because of this issue).
```

    meta.helm.sh/release-namespace: cattle-monitoring-system
    objectset.rio.cattle.io/id: default-mcc-rancher-monitoring-crd-cattle-fleet-local-system   # This required item is not in the above object.
  creationTimestamp: "2024-04-03T10:23:55Z"
  labels:
    app: rancher-monitoring-crd-manager
    app.kubernetes.io/managed-by: Helm
    objectset.rio.cattle.io/hash: 2da503261617e9ea2da822d2da7cdcfccad847a9    # This required item is not in the above object.
  name: rancher-monitoring-crd-manager
...
rules:
- apiGroups:
  - apiextensions.k8s.io
  resources:
  - customresourcedefinitions
  verbs:
  - create
  - get
  - patch
  - delete
  - update
```

### Workaround

1. Patch ClusterRole object `rancher-monitoring-crd-manager` to add the `update` operation.

    ```
    $ cat > patchrules.yaml << EOF
    rules:
    - apiGroups:
      - apiextensions.k8s.io
      resources:
      - customresourcedefinitions
      verbs:
      - create
      - get
      - patch
      - delete
      - update
    EOF

    $ kubectl patch ClusterRole rancher-monitoring-crd-manager --patch-file ./patchrules.yaml --type merge

    $ rm ./patchrules.yaml
    ```

> 1. Patch the missing objects with the required annotations and labels.

Reviewer: ```suggestion
1. Patch the `rancher-monitoring-crd-manager` ClusterRole object to add the `update` operation.
```

Author: Update accordingly with a bit change: there 4 objects to patch in this step, hence prfer `the missing objects` than a specific object; and this step is for `annotations and labels`.

    ```
    $ cat > patchhash.yaml << EOF
    metadata:
      annotations:
        objectset.rio.cattle.io/id: default-mcc-rancher-monitoring-crd-cattle-fleet-local-system
      labels:

>         objectset.rio.cattle.io/hash: 2da503261617e9ea2da822d2da7cdcfccad847a9

Reviewer 2: If we could explain how this hash value is calculated, users who are applying the patch would be more confident. Is this value always the same for every cluster?

Author: Explained  https://github.com/harvester/docs/pull/541#issue-2223364760

Reviewer 2: Oh cool, didn't noticed that. Thanks for the explanation.

    EOF

    $ kubectl patch ClusterRole rancher-monitoring-crd-manager --patch-file ./patchhash.yaml --type merge

    $ kubectl patch ClusterRoleBinding rancher-monitoring-crd-manager --patch-file ./patchhash.yaml --type merge

    $ kubectl patch ServiceAccount rancher-monitoring-crd-manager -n cattle-monitoring-system --patch-file ./patchhash.yaml --type merge

    $ kubectl patch ConfigMap rancher-monitoring-crd-manifest -n cattle-monitoring-system --patch-file ./patchhash.yaml --type merge

    $ rm ./patchhash.yaml
    ```

1. Check the `rancher-monitoring-crd` ManagedChart object.

>     After a few seconds the `rancher-monitoring-crd` ManagedChart object changes to ready.

Reviewer: ```suggestion
    After a few seconds, the status of the `rancher-monitoring-crd` ManagedChart object changes to `Ready`.
```

    ```
    $ kubectl get managedchart -n fleet-local rancher-monitoring-crd -oyaml
    apiVersion: management.cattle.io/v3
    kind: ManagedChart
    metadata:
    ...
      name: rancher-monitoring-crd
      namespace: fleet-local
    ...
    status:
      conditions:
      - lastUpdateTime: "2024-04-22T21:41:44Z"
        status: "True"
        type: Ready
    ...
    ```

>     The downstream objects also have no more complains.

Reviewer: ```suggestion
    Also, error indicators are no longer displayed for the downstream objects.
```

    ```
    ssh to Harvester VIP, then $ curl http://10.53.146.66/charts/index.yaml | grep "rancher-monitoring"
    ...
    name: rancher-monitoring
        - rancher-monitoring-102.0.0+up40.1.2.tgz
      rancher-monitoring-crd:
          catalog.cattle.io/release-name: rancher-monitoring-crd
        description: Installs the CRDs for rancher-monitoring.
        name: rancher-monitoring-crd
        - rancher-monitoring-crd-102.0.0+up40.1.2.tgz

    The chart version is 102.0.0+up40.1.2.
    ```

    The chart version must be identical to the value of the `version` field in the `rancher-monitoring-crd` ManagedChart object.

    :::note

    If those two versions are different, you need to stop and ask help from Harvester support or [community](https://github.com/harvester/harvester?tab=readme-ov-file#community).

    :::

1. Disable the `rancher-monitoring` [addon](../advanced/addons.md). Wait until the state changes to `Disabled`.

1. Delete the `rancher-monitoring-crd` ManagedChart.

   ```
   $ kubectl delete managedchart -n fleet-local rancher-monitoring-crd
   ```

1. Manually delete all objects marked as `missing`..

> 

Reviewer 2: Oh I meant users might have been enabled the `rancher-monitoring` add-on and created some custom resources like alertmanagerconfigs. Will they stay intact after the `rancher-monitoring-crd` ManagedChart was removed? If the answer is no, then it might be necessary to remind users to backup those custom resources first.

> 

Author: A good point, I will double check and validate, thanks.

> 

Author: Based on this important comment, we rebased the workaroud to patch the missing annotation and label manually, without deleting the managedchart and recreating it.
